irksome/stepper.py

```diff
@@ -265,3 +297,253 @@ def advance(self):
+    :arg KP: Proportional gain for step-size controller.
+    :arg max_reject: Maximum number of rejected timesteps in a row that
+            does not lead to a failure
+    :arg onscale_factor: Allowable tolerance in determining initial
+            timestep to be "on scale"
+    :arg safety_factor: Safety factor used when shrinking timestep if
+            a proposed step is rejected
+    :arg bcs: An iterable of :class:`firedrake.DirichletBC` containing
+            the strongly-enforced boundary conditions.  Irksome will
+            manipulate these to obtain boundary conditions for each
+            stage of the RK method.
+    :arg solver_parameters: A :class:`dict` of solver parameters that
+            will be used in solving the algebraic problem associated
+            with each time step.
+    :arg nullspace: A list of tuples of the form (index, VSB) where
+            index is an index into the function space associated with
+            `u` and VSB is a :class: `firedrake.VectorSpaceBasis`
+            instance to be passed to a
+            `firedrake.MixedVectorSpaceBasis` over the larger space
+            associated with the Runge-Kutta method
+    """
+    def __init__(self, F, butcher_tableau, t, dt, u0,
+                 bcs=None, appctx=None, solver_parameters=None,
+                 bc_type="DAE", splitting=AI, nullspace=None,
+                 tol=1.e-3, dtmin=1.e-15, dtmax=1.0, KI=1/15, KP=0.13,
+                 max_reject=10, onscale_factor=1.2, safety_factor=0.9):
+        assert butcher_tableau.btilde is not None
+        super(AdaptiveTimeStepper, self).__init__(F, butcher_tableau,
+                                                  t, dt, u0, bcs=bcs, appctx=appctx, solver_parameters=solver_parameters,
+                                                  bc_type=bc_type, splitting=splitting, nullspace=nullspace)
+
+        from firedrake.petsc import PETSc
+        self.print = lambda x: PETSc.Sys.Print(x)
+
+        self.dt_min = dtmin
+        self.dt_max = dtmax
+        self.dt_old = 0.0
+
+        self.delb = butcher_tableau.btilde - butcher_tableau.b
+        self.gamma0 = butcher_tableau.gamma0
+        self.KI = KI
+        self.KP = KP
+        self.max_reject = max_reject
+        self.onscale_factor = onscale_factor
+        self.safety_factor = safety_factor
+
+        self.error_func = Function(u0.function_space())
+        self.tol = tol
+        self.err_old = 0.0
+        self.contreject = 0
+
+        split_form = extract_terms(F)
+        self.dtless_form = -split_form.remainder
+
+        # Set up and cache boundary conditions for error estimate
+        if self.gamma0 != 0:
+            # Grab spaces for BCs
+            v = F.arguments()[0]
+            V = v.function_space()
+            num_fields = len(V)
+            num_stages = butcher_tableau.num_stages
+            btilde = butcher_tableau.btilde
+            ws = self.ws
+
+            class EmbeddedBCData(object):
```

Comment:
Is there any way we get code re-use from some other place where I've splatted out this wretched mess?

Reply:
I completely agree that we're recoding the same trick in several cases now.  This might be the worst place to look for better reuse, since it applies only to the case where we have an explicit first stage in an embedded scheme, which is not a case we have coded up elsewhere.